File: ruby/nucleotide-count/nucleotide_count.rb
class Nucleotide
  EMPTY_HISTOGRAM = "ACGT".chars.product([0]).to_h

  attr_reader :histogram

  def self.from_dna(seq)
    raise ArgumentError, "Invalid DNA sequence: #{seq}" unless /^[ACGT]*$/ =~ seq
    
    histogram = seq.chars.sort.chunk_while { |a, b| a == b }.map do |nucleotides|
      [nucleotides.first, nucleotides.size]
    end.to_h
    new(EMPTY_HISTOGRAM.merge(histogram))
  end

  def initialize(histogram)
    @histogram = histogram
  end

  def count(nucleotide)
    histogram[nucleotide]
  end
end

File: julia/nucleotide-count/nucleotide-count.jl
"""
    count_nucleotides(strand)

The count of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""
function count_nucleotides(strand)
    counts = Dict(n => 0 for n in "ACGT")
    for n in strand
        haskey(counts, n) || throw(DomainError(strand, "Invalid DNA strand: $strand"))
        counts[n] += 1
    end
    counts
end
